src/Donations/Repositories/DonationRepository.php

```diff
@@ -0,0 +1,519 @@
+
+        $renewals = $this->prepareQuery()
+            ->where('post_type', 'give_payment')
+            ->where('post_status', 'give_subscription')
+            ->whereIn('ID', function (QueryBuilder $builder) use ($subscriptionId) {
+                $builder
+                    ->select('donation_id')
+                    ->from('give_donationmeta')
+                    ->where('meta_key', DonationMetaKeys::SUBSCRIPTION_ID)
+                    ->where('meta_value', $subscriptionId);
+            });
+
+        return $renewals->orWhere('ID', $initialDonationId)->orderBy('post_date', 'DESC');
+    }
+
+    /**
+     * @unreleased
+     *
+     * @param  int  $donorId
+     *
+     * @return QueryBuilder
+     */
+    public function queryByDonorId($donorId)
+    {
+        return $this->prepareQuery()
+            ->where('post_type', 'give_payment')
+            ->whereIn('ID', function (QueryBuilder $builder) use ($donorId) {
+                $builder
+                    ->select('donation_id')
+                    ->from('give_donationmeta')
+                    ->where('meta_key', DonationMetaKeys::DONOR_ID)
+                    ->where('meta_value', $donorId);
+            })
+            ->orderBy('post_date', 'DESC');
+    }
+
+    /**
+     * @unreleased
+     *
+     * @param  Donation  $donation
+     *
+     * @return Donation
+     * @throws Exception|InvalidArgumentException
+     */
+    public function insert(Donation $donation)
+    {
+        $this->validateDonation($donation);
+
+        Hooks::dispatch('give_donation_creating', $donation);
+
+        $date = $donation->createdAt ? Temporal::getFormattedDateTime(
+            $donation->createdAt
+        ) : Temporal::getCurrentFormattedDateForDatabase();
+
+
+        DB::query('START TRANSACTION');
+
+        try {
+            DB::table('posts')
+                ->insert([
+                    'post_date' => $date,
+                    'post_date_gmt' => get_gmt_from_date($date),
+                    'post_modified' => $date,
+                    'post_modified_gmt' => get_gmt_from_date($date),
+                    'post_status' => $donation->status->getValue(),
+                    'post_type' => 'give_payment',
+                    'post_parent' => isset($donation->parentId) ? $donation->parentId : 0
+                ]);
+
+            $donationId = DB::last_insert_id();
+
+            foreach ($this->getCoreDonationMetaForDatabase($donation) as $metaKey => $metaValue) {
+                DB::table('give_donationmeta')
+                    ->insert([
+                        'donation_id' => $donationId,
+                        'meta_key' => $metaKey,
+                        'meta_value' => $metaValue,
+                    ]);
+            }
+        } catch (Exception $exception) {
+            DB::query('ROLLBACK');
+
+            Log::error('Failed creating a donation', compact('donation'));
```

Comment:
@jonwaldstein Is that right? I thought the second parameter was an associative array with expected keys (such as `'context'`)? I could be wrong.

Reply:
Yes this is right, context is just any array key that you want to add for additional information. 
<img width="651" alt="Screen Shot 2022-03-15 at 1 10 43 PM" src="https://user-images.githubusercontent.com/10138447/158433344-284ff7cc-3dac-4adf-a5dc-1c27de3090de.png">